bmq-sdk/src/main/java/com/bloomberg/bmq/Subscription.java

```diff
@@ -0,0 +1,332 @@
+     * Default maximum accumulated bytes of all outstanding messages that can be sent by the broker
+     * without being confirmed.
+     */
+    public static final long k_MAX_UNCONFIRMED_BYTES_DEFAULT = 33554432;
+
+    /** Default priority of a consumer with respect to delivery of messages. */
+    public static final int k_CONSUMER_PRIORITY_DEFAULT = 0;
+
+    private final Optional<Long> maxUnconfirmedMessages;
+    private final Optional<Long> maxUnconfirmedBytes;
+    private final Optional<Integer> consumerPriority;
+    private final SubscriptionExpression expression;
+    private final CorrelationId correlationId;
+
+    private Subscription(Builder builder) {
+        maxUnconfirmedMessages = builder.maxUnconfirmedMessages;
+        maxUnconfirmedBytes = builder.maxUnconfirmedBytes;
+        consumerPriority = builder.consumerPriority;
+        expression = builder.expression.orElse(new SubscriptionExpression());
+
+        if (builder.correlationId.isPresent()) {
+            correlationId = builder.correlationId.get();
+        } else {
+            correlationId = CorrelationIdImpl.nextId(builder.userData.orElse(null));
+        }
+    }
+
+    private Subscription() {
+        maxUnconfirmedMessages = Optional.empty();
+        maxUnconfirmedBytes = Optional.empty();
+        consumerPriority = Optional.empty();
+        expression = new SubscriptionExpression();
+        correlationId = CorrelationIdImpl.nextId();
+    }
+
+    /**
+     * Returns true if this {@code Subscription} is equal to the specified {@code obj}, false
+     * otherwise. Two {@code Subscription} are equal if they have equal properties.
+     *
+     * @param obj {@code Subscription} object to compare with
+     * @return boolean true if {@code Subscription} are equal, false otherwise
+     */
+    @Override
+    public boolean equals(Object obj) {
+
+        if (obj == null) return false;
+        if (obj == this) return true;
+        if (!(obj instanceof Subscription)) return false;
+        Subscription subscription = (Subscription) obj;
+
+        return getConsumerPriority() == subscription.getConsumerPriority()
+                && getMaxUnconfirmedBytes() == subscription.getMaxUnconfirmedBytes()
+                && getMaxUnconfirmedMessages() == subscription.getMaxUnconfirmedMessages()
+                && getExpression().equals(subscription.getExpression());
+    }
+
+    /**
+     * Returns a hash code value for this {@code Subscription} object.
+     *
+     * @return int a hash code value for this object.
+     */
+    @Override
+    public int hashCode() {
+
+        Long hash = (long) Long.hashCode(getMaxUnconfirmedMessages());
+        hash <<= Integer.SIZE;
+        hash |= (long) Long.hashCode(getMaxUnconfirmedBytes());
+        hash = (long) hash.hashCode();
+        hash <<= Integer.SIZE;
+        hash |= (long) Integer.hashCode(getConsumerPriority());
+        hash <<= Integer.SIZE;
+        hash |= (long) getExpression().hashCode();
```

Comment:
Same with correlationId

Reply:
Same, would rather not use CorrelationId to compare subscriptions